opm/simulators/flow/RSTConv.cpp

```diff
@@ -0,0 +1,130 @@
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <config.h>
+#include <opm/simulators/flow/RSTConv.hpp>
+
+#include <opm/input/eclipse/Schedule/Schedule.hpp>
+#include <opm/input/eclipse/Schedule/RSTConfig.hpp>
+
+#include <dune/common/fvector.hh>
+#include <dune/istl/bvector.hh>
+
+#include <algorithm>
+#include <cmath>
+
+namespace Opm {
+
+void RSTConv::init(const std::size_t numCells,
+                   const RSTConfig& rst_config,
+                   const std::array<int,3>& compIdx)
+{
+    const auto kw = rst_config.keywords.find("CONV");
+    if (kw == rst_config.keywords.end()) {
+        N_ = 0;
+        cnv_X_.clear();
+        return;
+    }
+
+    N_ = kw->second;
+    compIdx_ = compIdx;
+
+    cnv_X_.resize(3);
+    for (std::size_t i = 0; i < 3; ++i) {
+        if (compIdx_[i] != -1) {
+            cnv_X_[i].resize(numCells);
+        }
+    }
+}
+
+template<class ResidualVector>
+void RSTConv::update(const ResidualVector& resid)
+{
+    if (N_ == 0) {
+        return;
+    }
+
+    std::vector<int> cix(resid.size());
+    std::iota(cix.begin(), cix.end(), 0);
+    for (std::size_t i = 0; i < cnv_X_.size(); ++i) {
+        if (cnv_X_[i].empty()) {
+            continue;
+        }
+        std::partial_sort(cix.begin(), cix.begin() + N_, cix.end(),
+                          [&resid,c = compIdx_[i]](const int a, const int b)
+                          { return std::abs(resid[a][c]) > std::abs(resid[b][c]); });
+
+        this->gatherAndAccumulate(cix, resid, i);
+    }
+}
+
+template<class ResidualVector>
+void RSTConv::gatherAndAccumulate(const std::vector<int>& lIdx,
+                                  const ResidualVector& resid, int comp)
+{
+    if (comm_.size() == 1) {
+        for (int n = 0; n < N_; ++n) {
+            ++cnv_X_[comp][lIdx[n]];
+        }
+        return;
+    }
+
+    std::vector<double> values(comm_.rank() == 0 ? comm_.size() * N_ : N_);
+    std::vector<int> gIdx(comm_.rank() == 0 ? comm_.size() * N_ : N_);
+    for (int i = 0; i < N_; ++i) {
+        values[i] = std::abs(resid[lIdx[i]][compIdx_[comp]]);
+        gIdx[i] = globalCell_[lIdx[i]];
+    }
+
+    comm_.gather(gIdx.data(), gIdx.data(), N_, 0);
+    comm_.gather(values.data(), values.data(), N_, 0);
+    if (comm_.rank() != 0) {
+        return;
+    }
+
+    std::vector<std::pair<int,double>> data;
+    data.reserve(gIdx.size());
+    for (std::size_t i = 0; i < gIdx.size(); ++i) {
+        data.emplace_back(gIdx[i], values[i]);
+    }
+
+    std::partial_sort(data.begin(), data.begin() + N_, data.end(),
+                      [](const auto& a, const auto& b) { return a.second > b.second; });
+
+    for (int n = 0; n < N_; ++n) {
+        ++cnv_X_[comp][data[n].first];
+    }
```

Comment:
okay. increased readability (of pair code ) was my reason for doing it like that since I had to add additional data structures in any case, but i am not in any way invested in it.